src/atomicskiplist.ml

```diff
@@ -0,0 +1,221 @@
+type 'a markable_reference = { node : 'a; marked : bool }
+(** markable reference: stores a reference to a node and has a field to specify if it is marked *)
+
+type 'a node = {
+  key : int;
+  height : int;
+  next : 'a node option markable_reference Atomic.t array;
+}
+
+type 'a t = { head : 'a node }
+
+let max_height = 10
+
+(** create_new_node: creates a new node with some value and height *)
+let create_new_node value height =
+  let next = Array.init (max_height + 1) (fun _ -> Atomic.make {node = None; marked = false;})
+  in
+  { key = value; height; next }
+
+(** create_dummy_node_array: Creates a new array with the different node for each index *)
+let create_dummy_node_array () =
+  let arr = Array.init (max_height + 1) (fun _ -> create_new_node Int.max_int max_height)
+  in
+  arr
+
+(** Get a random level from 1 till max_height (both included) *)
+let get_random_level () =
+  let rec count_level cur_level =
+    if cur_level == max_height || Random.float 1.0 <= 0.5 then cur_level
+    else count_level (cur_level + 1)
+  in
+  count_level 1
+
+(** Create a new skiplist *)
+let create () =
+  let tail =
+    {
+      key = Int.max_int;
+      height = max_height;
+      next =
+        Array.make (max_height + 1)
+          (Atomic.make { node = None; marked = false });
+    }
+  in
+  let next = Array.init (max_height + 1) (fun _ -> Atomic.make { node = Some tail; marked=false; }) 
+  in
+  let head = { key = Int.min_int; height = max_height; next } in
+  { head }
+
+(** get_mark_ref: Returns the node and the mark from an Atomic markablereference *)
+let get_mark_ref atomic_ref =
+  let ref = Atomic.get atomic_ref in
+  (ref.node, ref.marked)
```

Comment:
It should be unnecessary to copy the values from the immutable record.  This whole helper function could be just removed.

Reply:
The Java implementation of `AtomicMarkableReference` includes functions like `getReference` and `get` which are analogous to `get_ref` and `get_mark_ref` in my implementation. In my aim to stick as close as possible to the implementation in "The Art of Multiprocessor Programming", I used similar functions.

Reply:
I guess that is a reasonable approach to get a working reference implementation.  However, Java and OCaml have many differences that make a direct transliteration from Java to OCaml undesirable.  For example, using options to emulate `null` is inefficient, because in OCaml that adds a level of indirection.  I would not recommend such an approach (i.e. direct transliteration) for a serious implementation.